internal/api/mfa.go

```diff
@@ -589,11 +589,23 @@ func (a *API) verifyPhoneFactor(w http.ResponseWriter, r *http.Request, params *
 		}
 		return unprocessableEntityError(ErrorCodeMFAChallengeExpired, "MFA challenge %v has expired, verify against another challenge or create a new challenge.", challenge.ID)
 	}
-	otpCode, shouldReEncrypt, err := challenge.GetOtpCode(config.Security.DBEncryption.DecryptionKeys, config.Security.DBEncryption.Encrypt, config.Security.DBEncryption.EncryptionKeyID)
-	if err != nil {
-		return internalServerError("Database error verifying MFA TOTP secret").WithInternalError(err)
+	var valid bool
+	var otpCode string
+	var shouldReEncrypt bool
+	var err error
+	if config.Sms.IsTwilioVerifyProvider() {
+		smsProvider, _ := sms_provider.GetSmsProvider(*config)
+		if err := smsProvider.(*sms_provider.TwilioVerifyProvider).VerifyOTP(factor.Phone.String(), nonce); err != nil {
+			return forbiddenError(ErrorCodeOTPExpired, "Token has expired or is invalid").WithInternalError(err)
+		}
+		valid = true
+	} else {
+		otpCode, shouldReEncrypt, err = challenge.GetOtpCode(config.Security.DBEncryption.DecryptionKeys, config.Security.DBEncryption.Encrypt, config.Security.DBEncryption.EncryptionKeyID)
```

Comment:
I think you are still working on this, but I did notice this code below could panic since the err is dropped in `smsProvider, _ := sms_provider.GetSmsProvider(*config)`. Maybe we could hoist that to the block above. Since it appears this is checking for an optional interface, we could add one:
```Go
type SmsProvider interface {
	SendMessage(phone, message, channel, otp string) (string, error)
}

type SmsProviderVerifier interface {
	 VerifyOTP(phone, code string) error
}
```

Then check the optional interface in below:
```suggestion
	smsProvider, err := sms_provider.GetSmsProvider(*config)
	if err == nil { ...	}
	if vr, ok := smsProvider.(sms_provider.SmsProviderVerifier); ok {
		if err := vr.VerifyOTP(factor.Phone.String(), code); err != nil {
			return forbiddenError(ErrorCodeOTPExpired,
				"Token has expired or is invalid").WithInternalError(err)
		}
		valid = true
	} else {
		otpCode, shouldReEncrypt, err = ...
```

Reply:
That's a good suggestion, thanks!

Yeah unfortunately haven't had time to close this one out, will do so after finishing MFA for WebAuthn hopefully